### ISSUE TITLE
 Esc key to cancel quickMove also clears Cmd-Shift-K search box...

### ISSUE THREAD

**Issue by fredstluka** [RealRaven2000/QuickFolders#41]
The Esc key cancels a quickMove operation, clearing the text typed
into the quickMove box and hiding the dropdown list of matches.
Good!  But, it also clears the native TB "Filter these messages"
box at the top right of the TB window that I opened via Cmd-Shift-K.

Perhaps QF can catch the Esc key when the quickMove box is open and 
prevent it from propagating to TB?  Then Esc would cancel the 
quickMove if one was pending, and would clear the  "Filter these 
messages" box otherwise.  In JavaScript Web applications, it's easy 
for each event handler to specify whether or not the event
(keystroke, mouse click, etc.) should propagate to other event 
handlers for other containing HTML elements.  Perhaps it's easy to 
do here also?

That's how the "Find in page" box currently works.  If I hit 
Cmd-Shift-K to see the "Filter these messages" box, and type 
something there, Esc clears what I typed.  But while the filter
text is still showing, if I type Cmd-F to see the "Find in page"
box at the bottom left of the message body pane, and type in 
some text to find, Esc hides the "Find in page" box, but does NOT
clear the "Filter these messages" box.

BTW, according to this old Bugzilla ticket:
- https://www.mozdev.org/bugs/show_bug.cgi?id=26660

the change to add this Esc key functionality in version 4.15 on 
May 21, 2019 was supposed to add it only for Shift-Esc, not for
Esc w/o the Shift key.  Was it coded wrong?  Or is there a 
pattern here caused by my use of Mac that unshifted keys do the
action intended for shifted keys?  See this other ticket where
unshifted "s" does the action of Shift-S:
- https://github.com/RealRaven2000/QuickFolders/issues/35

But note that this problem does not occur for all unshifted keys.
For example, unshifted "m" still toggles the native TB read/unread 
status of a message, while Shift-M starts a QF quickMove operation.

In any case, I REALLY like the Esc key as a way to cancel a 
quickMove.  It is very consistent with other apps I've used on 
Mac, Windows, Unix, Linux, VMS, etc. for decades.  I started 
using it naturally without having to read it in any docs.
But it would NEVER have occurred to me to use "Shift-Esc".

So, if it's possible to fix this bug, but continue to use 
unshifted Esc, I would prefer that over changing to Shift-Esc.

QF version 4.18.1
TB version 68.9.0
Mac OS X version 10.11.6
--Fred

**Comment by fredstluka** [comment]
BTW, the Premium Features page:
- http://quickfolders.org/premium.html

says:
- quickMove
  - Hit ESC to hide the panel.
- quickJump
  - hit ESC to cancel.  

So, perhaps it WAS originally intended to work unshifted.

--Fred

**Comment by RealRaven2000** [comment]
> The Esc key cancels a quickMove operation, clearing the text typed
> into the quickMove box and hiding the dropdown list of matches.
> Good! But, it also clears the native TB "Filter these messages"
> box at the top right of the TB window that I opened via Cmd-Shift-K.

oh, ok. I assume this is while the quickMove box has focus? I probably need to consume the event better and make sure it doesn't propagate...


> > Perhaps QF can catch the Esc key when the quickMove box is open and
> prevent it from propagating to TB? 
+1

> Then Esc would cancel the
> quickMove if one was pending, and would clear the "Filter these
> messages" box otherwise. In JavaScript Web applications, it's easy
> for each event handler to specify whether or not the event
> (keystroke, mouse click, etc.) should propagate to other event
> handlers for other containing HTML elements. Perhaps it's easy to
> do here also?
same thing, events do bubble up (and down) but it can be prevented  with some JS.


> BTW, according to this old Bugzilla ticket:
> 
>     * https://www.mozdev.org/bugs/show_bug.cgi?id=26660
> 
> 
> the change to add this Esc key functionality in version 4.15 on
> May 21, 2019 was supposed to add it only for Shift-Esc, not for
> Esc w/o the Shift key. Was it coded wrong? 
I think I simplified it later, because technically Shift+Esc is a "double keystroke", and that's just harder to type and remember. not sure which ticket handles that.

> But note that this problem does not occur for all unshifted keys.
> For example, unshifted "m" still toggles the native TB read/unread
> status of a message, while Shift-M starts a QF quickMove operation.
Like I said, it's intended behavior and may be documented in a later bugzilla bug. I won't really be able to clean up the old tickets as the mozdev server is going away, and we are starting afresh on github.

> So, if it's possible to fix this bug, but continue to use
> unshifted Esc, I would prefer that over changing to Shift-Esc.
You already got it. It's not a bug, it's a feature. And may even be documented somewhere. 

Also see:
http://quickfolders.org/version.html#quickJump

when you are using quickMove "The Escape key will hide the red box temporarily." so it doesn't cancel moving multiple mails. This way you can add more mails to the next quickMove operation until you decide where the target folder is.

**Comment by RealRaven2000** [comment]
> 
> 
> BTW, the Premium Features page:
> 
>     * http://quickfolders.org/premium.html
> 
> 
> says:
> 
>     * quickMove
>       
>       * Hit ESC to hide the panel.
> 
>     * quickJump
>       
>       * hit ESC to cancel.
> 
> 
> So, perhaps it WAS originally intended to work unshifted.

I think it was streamlined and then documented on premium.html "after the fact". Never mind the old bugzilla tickets :)

**Comment by fredstluka** [comment]
> oh, ok. I assume this is while the quickMove box has focus? 

Yes.

> I probably need to consume the event better and make sure it doesn't propagate...

Yes, exactly.

 
> same thing, events do bubble up (and down) but it can be prevented  with some JS.

Right.  In all browsers these days, you can just return false from the event procedure.  Used to have to do stuff like:

```javascript
    // Note: None of these are needed for Firefox 2.0.0.8 or IE 6.0.2800.1106.
    //       Returning false is sufficient.
    // if (typeof (evt.returnValue) != "undefined")
    // {
    //     evt.returnValue = false;
    // }
    // if (typeof (evt.cancelBubble) != "undefined")
    // {
    //     evt.cancelBubble = true;
    // }
    // if (typeof (evt.stopPropagation) != "undefined")
    // {
    //     evt.stopPropagation();
    // }
    // if (typeof (evt.keyCode) != "undefined")
    // {
    //     evt.keyCode = 0;
    // }
```

I assume it works similarly for TB addons?


> I think I simplified it later, because technically Shift+Esc is a "double keystroke", and that's just harder to type and remember. not sure which ticket handles that.

OK.  Good!  I prefer unshifted Esc anyhow.

> Like I said, it's intended behavior and may be documented in a later bugzilla bug. I won't really be able to clean up the old tickets as the mozdev server is going away, and we are starting afresh on github.

Don't worry about it.  Move forward, adding new fixes/features and documenting them.  No need to go back and edit comments in old tickets.

--Fred

**Comment by fredstluka** [comment]
BTW, when you're inserting lines into my quoted text using Markdown here, always leave a blank line before your insertion.  Otherwise, Markdown shows your insertion as part of my quoted text.

--Fred

**Comment by RealRaven2000** [comment]
I just added `event.preventDefault()` to the routine `QuickFolders.Interface,findFolderKeyPress` that appears to have fixed it. unsurprisingly I already used it for other key functions within the same place. Easy fix.

Here is a test version:
[QuickFolders-wx-4.18.2pre2.zip](https://github.com/RealRaven2000/QuickFolders/files/4879190/QuickFolders-wx-4.18.2pre2.zip)

**Comment by RealRaven2000** [comment]
fixed in 4.18.2 and 5.0

**Comment by fredstluka** [comment]
Yes, it works fine now (in 5.1).  Thanks!  --Fred
